src/legacy/core_plugins/data/public/filter/filter_manager/filter_state_manager.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Filter, FilterStateStore } from '@kbn/es-query';
+
+import _ from 'lodash';
+import { State } from 'ui/state_management/state';
+import { FilterManager } from './filter_manager';
+
+/**
+ * FilterStateManager is responsible for watching for filter changes
+ * and synching with FilterManager, as well as syncing FilterManager changes
+ * back to the URL.
+ **/
+export class FilterStateManager {
+  filterManager: FilterManager;
+  globalState: State;
+  getAppState: () => State;
+  prevGlobalFilters: Filter[] | undefined;
+  prevAppFilters: Filter[] | undefined;
+  interval: NodeJS.Timeout | undefined;
+
+  constructor(globalState: State, getAppState: () => State, filterManager: FilterManager) {
+    this.getAppState = getAppState;
+    this.globalState = globalState;
+    this.filterManager = filterManager;
+
+    this.watchFilterState();
+
+    this.filterManager.getUpdates$().subscribe(() => {
+      this.updateAppState();
+    });
+  }
+
+  destroy() {
+    if (this.interval) {
+      clearInterval(this.interval);
+    }
+  }
+
+  private watchFilterState() {
+    // This is a temporary solution to remove rootscope.
+    // Moving forward, state should provide observable subscriptions.
+    this.interval = setInterval(() => {
```

Comment:
what about 
```
const appState = this.getAppState();
appState.on('fetch_with_changes', handleUpdatesFromUrl);
this.globalState.on('fetch_with_changes', handleUpdatesFromUrl);
```
instead of interval ?